src/mumble/widgets/TrayIcon.cpp

```diff
@@ -0,0 +1,208 @@
+	QObject::connect(m_contextMenu, &QMenu::aboutToShow, this, &TrayIcon::updateContextMenu);
+
+	// Some window managers hate it when a tray icon sets an empty context menu...
+	updateContextMenu();
+
+	setContextMenu(m_contextMenu);
+
+	show();
+}
+
+void TrayIcon::updateIcon() {
+	QIcon *newIcon = nullptr;
+
+	ClientUser *p = ClientUser::get(Global::get().uiSession);
+
+	if (Global::get().s.bDeaf) {
+		newIcon = &Global::get().mw->qiIconDeafSelf;
+	} else if (p && p->bDeaf) {
+		newIcon = &Global::get().mw->qiIconDeafServer;
+	} else if (Global::get().s.bMute) {
+		newIcon = &Global::get().mw->qiIconMuteSelf;
+	} else if (p && p->bMute) {
+		newIcon = &Global::get().mw->qiIconMuteServer;
+	} else if (p && p->bSuppress) {
+		newIcon = &Global::get().mw->qiIconMuteSuppressed;
+	} else if (Global::get().s.bStateInTray && Global::get().bPushToMute) {
+		newIcon = &Global::get().mw->qiIconMutePushToMute;
+	} else if (p && Global::get().s.bStateInTray) {
+		switch (p->tsState) {
+			case Settings::Talking:
+			case Settings::MutedTalking:
+				newIcon = &Global::get().mw->qiTalkingOn;
+				break;
+			case Settings::Whispering:
+				newIcon = &Global::get().mw->qiTalkingWhisper;
+				break;
+			case Settings::Shouting:
+				newIcon = &Global::get().mw->qiTalkingShout;
+				break;
+			case Settings::Passive:
+			default:
+				newIcon = &Global::get().mw->qiTalkingOff;
+				break;
+		}
+	} else {
+		newIcon = &Global::get().mw->qiIcon;
+	}
+
+	m_iconLock.lockForRead();
+	bool isStatusIcon = newIcon == m_statusIcon;
+	m_iconLock.unlock();
+
+	if (!isStatusIcon) {
+		m_iconLock.lockForWrite();
+		m_statusIcon = newIcon;
+		m_iconLock.unlock();
+
+		applyIcon(m_statusIcon);
+	}
+}
+
+void TrayIcon::on_icon_clicked(QSystemTrayIcon::ActivationReason reason) {
+	switch (reason) {
+		case QSystemTrayIcon::Trigger:
+#ifndef Q_OS_MAC
+			// macOS is special as it both shows the context menu AND triggers the action.
+			// We only want at most one of those and since we can not prevent showing
+			// the menu, we skip the action.
+			toggleShowHide();
+#endif
+			break;
+		case QSystemTrayIcon::Unknown:
+		case QSystemTrayIcon::Context:
+		case QSystemTrayIcon::DoubleClick:
+		case QSystemTrayIcon::MiddleClick:
+		default:
+			break;
+	}
+}
+
+void TrayIcon::updateContextMenu() {
+	m_contextMenu->clear();
+
+	if (Global::get().mw->isVisible() && !Global::get().mw->isMinimized()) {
+		m_hideAction->setEnabled(QSystemTrayIcon::isSystemTrayAvailable());
+		m_contextMenu->addAction(m_hideAction);
+	} else {
+		m_contextMenu->addAction(m_showAction);
+	}
+
+	m_contextMenu->addSeparator();
+
+	m_contextMenu->addAction(Global::get().mw->qaAudioMute);
+	m_contextMenu->addAction(Global::get().mw->qaAudioDeaf);
+	m_contextMenu->addAction(Global::get().mw->qaTalkingUIToggle);
+	m_contextMenu->addSeparator();
+	m_contextMenu->addAction(Global::get().mw->qaQuit);
+}
+
+void TrayIcon::applyIcon(QIcon *icon) {
+	QWriteLocker locker(&m_iconLock);
+
+	setIcon(*icon);
```

Comment:
```suggestion
void TrayIcon::applyIcon(QIcon &icon) {
	QWriteLocker locker(&m_iconLock);

	setIcon(&icon);
```
also, can the passed icon be `const`? If `setIcon` needs a copy of the icon, this function should take the icon by value and `move` it into `setIcon`.

Reply:
I need to use pointers, because ``QIcon`` (and by extension ``QSystemTrayIcon``) does not implement a way to check for equality. And I need the equality check for the blinking icon in ``m_highlightTimer``

Reply:
You can still get the address of the original thing through its reference via the `&` operator.

Reply:
Ah :bulb: Updated to use as little pointers as possible.